controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -602,6 +613,446 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
+                result.complete(true);
+            }
+
+            @Override
+            public void segmentDeleted(WireCommands.SegmentDeleted segmentDeleted) {
+                log.info(requestId, "deleteTableSegment {} SegmentDeleted.", qualifiedName);
+                result.complete(true);
+            }
+
+            @Override
+            public void tableSegmentNotEmpty(WireCommands.TableSegmentNotEmpty tableSegmentNotEmpty) {
+                log.warn(requestId, "deleteTableSegment {} TableSegmentNotEmpty.", qualifiedName);
+                result.completeExceptionally(new WireCommandFailedException(type, WireCommandFailedException.Reason.TableSegmentNotEmpty));
+            }
+
+            @Override
+            public void processingFailure(Exception error) {
+                log.error(requestId, "deleteTableSegment {} failed.", qualifiedName, error);
+                result.completeExceptionally(error);
+            }
+
+            @Override
+            public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenCheckFailed) {
+                result.completeExceptionally(
+                        new WireCommandFailedException(new AuthenticationException(authTokenCheckFailed.toString()),
+                                                       type, WireCommandFailedException.Reason.AuthFailed));
+            }
+        };
+
+        WireCommands.DeleteTableSegment request = new WireCommands.DeleteTableSegment(requestId, qualifiedName, mustBeEmpty, delegationToken);
+        sendRequestAsync(request, replyProcessor, result, clientCF, ModelHelper.encode(uri));
+        return result;
+    }
+
+    /**
+     * This method sends a WireCommand to update table entries.
+     *
+     * @param scope               Stream scope.
+     * @param stream              Stream name.
+     * @param segmentId           Id of the segment to be created.
+     * @param entries             List of {@link TableEntry}s to be updated.
+     * @param hostControllerStore Host controller store.
+     * @param clientCF            Client connection factory.
+     * @param delegationToken     The token to be presented to the segmentstore.
+     * @param clientRequestId     Request id.
+     * @return A CompletableFuture that, when completed normally, will contain the current versions of each {@link TableEntry}
+     * If the operation failed, the future will be failed with the causing exception. If the exception is retryable then the future
+     * will be failed with {@link WireCommandFailedException}.
+     */
+    public CompletableFuture<List<KeyVersion>> updateTableEntries(final String scope,
+                                                                  final String stream,
+                                                                  final long segmentId,
+                                                                  final List<TableEntry<byte[], byte[]>> entries,
```

Comment:
This method takes TableEntry and the removeKeys method takes TableKey as input. 
This is rather unintuitive. it is better to have both follow the same pattern so that we are either using tableKey for versions or TableEntry for versions. 
Ideally if we can decouple the KeyVersion from the entry and use that it would be best.

Reply:
`TableEntry` does have a `getKey` on it. You can use it to fetch the version and determine if this is a conditional update or not.